src/moonlink_connectors/src/rest_ingest/schema_util.rs

```diff
@@ -0,0 +1,359 @@
+}
+
+impl FieldSchemaBuilder {
+    #[must_use]
+    pub fn new(name: String, data_type: String, nullable: bool) -> Self {
+        Self {
+            name,
+            data_type,
+            nullable,
+            fields: None,
+            item: None,
+        }
+    }
+
+    pub fn name(mut self, name: impl Into<String>) -> Self {
+        self.name = name.into();
+        self
+    }
+
+    pub fn data_type(mut self, data_type: impl Into<String>) -> Self {
+        self.data_type = data_type.into();
+        self
+    }
+
+    pub fn nullable(mut self, nullable: bool) -> Self {
+        self.nullable = nullable;
+        self
+    }
+
+    pub fn fields(mut self, fields: Option<Vec<FieldSchemaBuilder>>) -> Self {
+        self.fields = fields;
+        self
+    }
+
+    pub fn item(mut self, item: Option<Box<FieldSchemaBuilder>>) -> Self {
+        self.item = item;
+        self
+    }
+
+    pub fn build(self) -> Result<FieldSchemaBuilder, SchemaBuildError> {
+        Ok(FieldSchemaBuilder {
+            name: self.name,
+            data_type: self.data_type,
+            nullable: self.nullable,
+            fields: self.fields,
+            item: self.item,
+        })
+    }
+
+    /// Parse a decimal type string in the form of
+    ///   - decimal(precision)
+    ///   - decimal(precision, scale)
+    fn parse_decimal(&self) -> Result<DecimalType, SchemaBuildError> {
+        let data_type_lowercase = self.data_type.to_lowercase();
+        let data_type_str = data_type_lowercase.as_str();
+        let inner = &data_type_str[8..data_type_str.len() - 1];
+        let parts: Vec<&str> = inner.split(',').map(|s| s.trim()).collect();
+        // Arrow type allows no "scale", which defaults to 0.
+        if parts.len() == 1 {
+            let precision: u8 = parts[0]
+                .parse()
+                .map_err(|_| SchemaBuildError::InvalidDecimal(data_type_str.to_string()))?;
+            Ok(DecimalType {
+                precision,
+                scale: 0,
+            })
+        } else if parts.len() == 2 {
+            // decimal(precision,
+            let precision: u8 = parts[0]
+                .parse()
+                .map_err(|_| SchemaBuildError::InvalidDecimal(data_type_str.to_string()))?;
+            let scale: i8 = parts[1]
+                .parse()
+                .map_err(|_| SchemaBuildError::InvalidDecimal(data_type_str.to_string()))?;
+            Ok(DecimalType { precision, scale })
+        } else {
+            Err(SchemaBuildError::InvalidDecimal(data_type_str.to_string()))
+        }
+    }
+
+    /// Build an Arrow `Field` from a `FieldSchema`.
+    ///
+    /// Returns an Arrow field and mutates the `field_id` as a side effect
+    /// Returns an error if the field schema is invalid.
+    fn build_field_from_schema(
+        &self,
+        override_name: Option<&str>,
+        field_id: &mut i32,
+    ) -> Result<Field, SchemaBuildError> {
```

Comment:
You might also need to refactor the error, essentially all errors should be a centralized error type in `crate::Error`, current implementation loses track of error source location